Checked out `4822e61` and ran it: `cargo fmt --check` clean, `cargo test --lib --bins` green (341 lib + 13 `llmman-bench`).

The diagnosis is right and I confirmed both halves independently.

**The CI target bug is real and worse than the description says.** `Cargo.toml` does have a `[lib]` target, so the workflow comment that `cargo test --bin llmman` exists "since this crate has no lib target" is simply stale. `src/main.rs` contains no `#[cfg(test)]` at all, so the `Unit tests` job has been running **zero tests on all five targets** while reporting green. `--lib --bins` picks up 341+13. That's the single most valuable thing in this PR.

**The E2E failure is real too.** Latest main run [33090803965](https://github.com/llmmanorg/llmman/actions/runs/33090803965): only the three Linux legs fail (x86_64 docker, x86_64 podman, aarch64 docker); macOS and Windows pass — exactly the procps-ng-vs-BSD split you'd predict. The log ends with:

```
[spawn_with_timeout] pid=12247 spawned: `llmman pull` against a daemon that dies at startup
##[error]The runner has received a shutdown signal.
test ensure_server_fails_fast_when_daemon_dies_at_startup ...
```

So: right bug, right root cause. But I don't think this branch should merge as-is, for four reasons.

---

### 1. It collides with #317, which fixes the same bug more completely

#317 (opened ~4h before this one) is the same kill(2) change. I trial-merged both: they **conflict in `src/daemon.rs`**, so one of them has to be rebased regardless.

More importantly, this PR only converts `signal_group`. `kill_daemon` (`src/daemon.rs:267`) still does:

```rust
let group = Command::new("kill").args([signal, &format!("-{pid}")]).status();
```

That is the identical procps-ng footgun, still live on this branch. And it's a *user-facing* path, not just a CI one: `kill_daemon` is reached from `stop_stale_daemon`, which runs whenever any command meets a daemon from an older install. On Ubuntu that means "upgrade llmman, run `llmman pull`, and SIGTERM every process your user owns" — then SIGKILL 5s later via the escalation. So this branch leaves the more damaging instance of the bug in place while fixing the one CI happens to trip over. #317 converts both.

### 2. No guard on the negative pid

```rust
unsafe { libc::kill(-(pid as libc::pid_t), signal) == 0 }
```

Two problems. `pid == 1` gives `kill(-1, sig)` — the exact user-wide broadcast this PR is fixing. `pid == 0` gives `kill(0, sig)` — the caller's own group. And `pid as libc::pid_t` is a lossy `u32 -> i32` cast: a value above `i32::MAX` wraps negative, and `-(negative)` then signals a single unrelated pid (or panics on `i32::MIN` in a debug build). Yes, `child.id()` won't realistically return those today, but this function is one refactor away from taking a pid parsed off the wire in `DaemonIdentity`, and the failure mode is "take down the machine". #317's `kill_group` has the `pid <= 1` and `i32::try_from` guards plus a unit test (`kill_group_refuses_pids_that_would_signal_everything`) that pins them with signal 0.

### 3. Nothing here is covered by a test that runs on a PR

`e2e` is gated on `github.event_name == 'push'`, so the only job that exercises `ensure_server_fails_fast_when_daemon_dies_at_startup` cannot run on this PR — it stays SKIPPED, as it is above. That means the daemon half of this change lands unverified and is first exercised on main.

The linked "working CI example from fork" doesn't close that gap: run [33123225590](https://github.com/onuralpszr/llmman/actions/runs/33123225590) is on `063acdea`, and I fetched that commit — it's upstream's `fix(ci): retry, verify, and hard-fail integration CLI installs (#304)`, with `src/daemon.rs` still calling `Command::new("kill")` and `ci.yml` still on `cargo test --bin llmman`. **None of this PR's changes are in that run.** It's evidence that E2E passed on an older commit, not that this branch fixes it.

The fix is unit tests that run in the `test` job. #317 already has the shape: spawn a `process_group(0)` target plus a control child in the harness's own group, assert the target dies and the control survives — a regression would fail an assertion instead of the runner.

### 4. Two changes are outside the stated scope

`DownloadMarker::touch` in `src/llama_release.rs` is not mentioned in the title, the description, or either commit message:

```rust
let _ = std::fs::write(path, b"");
if let Ok(file) = std::fs::File::options().write(true).open(path) {
    let _ = file.set_modified(std::time::SystemTime::now());
}
```

The premise ("writing zero bytes leaves the mtime alone on Windows") may well be right, but it's an untested claim about a platform, it's a second `open()` where `File::create(path)` + `set_modified` would do, and it silently changes when `download_in_progress()` reports a live download — which is what decides whether a 60s startup timeout kills the daemon or leaves it running. That deserves its own PR with a test against `marker_is_fresh`.

The `has_exited` / `owns_group` rework is at least related, but it's a behavioural change to startup failure handling riding along with a CI fix, and CodeRabbit's point on it stands: `waitid` returning `-1` is folded into `Ok(false)`, so a failed status check makes `ensure_server` sit out the full 60s timeout instead of reporting. `EINTR` needs a retry, anything else needs `last_os_error()`.

---

### What I'd like to see

**Split this into two PRs.**

- **PR A — the CI target fix only.** `.github/workflows/ci.yml`: `--lib --bins` plus the corrected comment. That is self-contained, self-evidently correct, immediately valuable, has zero overlap with #317, and I'll merge it as soon as it's up. Please also fix the stale "this crate has no lib target — see Cargo.toml" line, since that claim is what caused the bug.
- **PR B — the daemon work.** Best outcome is landing #317 (it converts both call sites and comes with tests) and rebasing whatever is genuinely additive here on top: the `has_exited`/`WNOWAIT` pid-reservation reasoning is a good catch that #317 doesn't have. If you'd rather carry it here instead, then this branch needs `kill_daemon` converted too, the `pid <= 1` / `try_from` guards, unit tests in the `test` job, and the `waitid` error propagated.
- **PR C — the `DownloadMarker::touch` mtime fix**, on its own, with a test.

Worth coordinating with @ericcurtin on #317 vs #320 before spending more time on the daemon half — no point doing the work twice.

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -167,8 +167,8 @@ jobs:
           if-no-files-found: error
 
   # --------------------------------------------------------------------------
-  # Unit tests: every #[cfg(test)] module under src/ (`cargo test --bin`,
-  # since this crate has no lib target — see Cargo.toml). These are pure-
+  # Unit tests: every #[cfg(test)] module under src/, which live in the
+  # lib target and the llmman-bench binary. These are pure-
   # Rust logic tests with no OS-specific behavior, but the crate itself
   # still links the Go shim via CGO, so a target-specific toolchain issue
   # (the exact kind the `build`/`e2e` jobs' own git history is full of —
@@ -288,8 +288,8 @@ jobs:
       # comment on cargo test compiling its own separate test binary,
       # which links against the same Go-shim-produced llmman_shim.lib the
       # main binary does and hit the exact same lld-link requirement.
-      - name: cargo test --bin llmman
-        run: cargo test --release --target ${{ matrix.target }} --bin llmman
+      - name: cargo test --lib --bins
+        run: cargo test --release --target ${{ matrix.target }} --lib --bins
         env:
           RUSTFLAGS: ${{ matrix.rustflags }}
 

--- a/src/llama_release.rs
+++ b/src/llama_release.rs
@@ -459,7 +459,12 @@ impl DownloadMarker {
     /// download from a crashed one whose Drop never ran.
     fn touch(&self) {
         if let Some(path) = &self.0 {
+            // Windows keeps the old mtime when a write is zero bytes,
+            // so set it explicitly.
             let _ = std::fs::write(path, b"");
+            if let Ok(file) = std::fs::File::options().write(true).open(path) {
+                let _ = file.set_modified(std::time::SystemTime::now());
+            }
         }
     }
 }

